### PR TITLE
Smart Contract deploy transactions parser

### PR DESCRIPTION
**Pull request by popenta**
Implemented the smart contract deploy transactions parser based on the sdk-specs.

### CODE REVIEW

File: multiversx_sdk/core/transactions_outcome_parsers/smart_contract_transactions_outcome_parser.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+from multiversx_sdk.core.address import Address
+from multiversx_sdk.core.transactions_outcome_parsers.resources import (
+    TransactionEvent, TransactionOutcome, find_events_by_identifier)
+from multiversx_sdk.core.transactions_outcome_parsers.smart_contract_transactions_outcome_parser_types import (
+    DeployedSmartContract, SmartContractDeployOutcome)
+from multiversx_sdk.network_providers.constants import DEFAULT_ADDRESS_HRP
+
+
+class SmartContractTransactionsOutcomeParser:
+    def __init__(self) -> None:
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/core/transactions_outcome_parsers/smart_contract_transactions_outcome_parser.py:10]: Yes, we'll implement that in the future. We need the `abi` for that but we currently don't have it in sdk-py.

File: multiversx_sdk/converters/transactions_converter_test.py [hunk 2]
```diff
@@ -47,3 +65,147 @@ def test_transaction_converter():
...
+    assert actual_tx_outcome.logs.events[0].topics == expected_tx_outcome.logs.events[0].topics
+
+    assert actual_tx_outcome.transaction_results[0].sender == expected_tx_outcome.transaction_results[0].sender
+    assert actual_tx_outcome.transaction_results[0].receiver == expected_tx_outcome.transaction_results[0].receiver
+    assert actual_tx_outcome.transaction_results[0].data == expected_tx_outcome.transaction_results[0].data
+    assert actual_tx_outcome.transaction_results[0].logs.address == expected_tx_outcome.transaction_results[0].logs.address
+    assert actual_tx_outcome.transaction_results[0].logs.events[0].address == expected_tx_outcome.transaction_results[0].logs.events[0].address
+    assert actual_tx_outcome.transaction_results[0].logs.events[0].identifier == expected_tx_outcome.transaction_results[0].logs.events[0].identifier
+    assert actual_tx_outcome.transaction_results[0].logs.events[0].data_items == expected_tx_outcome.transaction_results[0].logs.events[0].data_items
+    assert actual_tx_outcome.transaction_results[0].logs.events[0].topics == expected_tx_outcome.transaction_results[0].logs.events[0].topics
```
Comment (andreibancioiu) [inline multiversx_sdk/converters/transactions_converter_test.py:148]: No way to do a shorter assert? Deep equality etc.?

File: multiversx_sdk/core/transactions_outcome_parsers/smart_contract_transactions_outcome_parser.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        contracts = [self._parse_sc_deploy_event(event) for event in events]
+
+        return SmartContractDeployOutcome(direct_call_outcome.return_code, direct_call_outcome.return_message, contracts)
+
+    def _parse_sc_deploy_event(self, event: TransactionEvent) -> DeployedSmartContract:
+        contract_address_topic = event.topics[0] if event.topics[0] else b''
+        owner_address_topic = event.topics[1] if event.topics[1] else b''
+        code_hash_topic = event.topics[2] if event.topics[2] else b''
+
+        contract_address = Address(contract_address_topic, DEFAULT_ADDRESS_HRP).to_bech32() if len(contract_address_topic) else ""
```
Comment (andreibancioiu) [inline multiversx_sdk/core/transactions_outcome_parsers/smart_contract_transactions_outcome_parser.py:25]: In the future, we have to parametrize such outcome parsers to receive the `hrp`.

File: multiversx_sdk/core/transactions_outcome_parsers/smart_contract_transactions_outcome_parser_test.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        tx_outcome = transaction_converter.transaction_on_network_to_outcome(tx_on_network)
+
+        parsed = self.parser.parse_deploy(tx_outcome)
+
+        assert parsed.return_code == ""
+        assert parsed.return_message == ""
+        assert len(parsed.contracts) == 0
+        assert parsed.contracts == []
+
+    @pytest.mark.networkInteraction
```
Comment (andreibancioiu) [inline multiversx_sdk/core/transactions_outcome_parsers/smart_contract_transactions_outcome_parser_test.py:124]: :+1:

File: multiversx_sdk/core/transactions_outcome_parsers/smart_contract_transactions_outcome_parser_test.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        tx_converter = TransactionsConverter()
+
+        tx_on_network = proxy.get_transaction(successful_tx_hash)
+        tx_outcome = tx_converter.transaction_on_network_to_outcome(tx_on_network)
+
+        parsed = self.parser.parse_deploy(tx_outcome)
+        assert parsed.contracts[0].address == "erd1qqqqqqqqqqqqqpgq29deu3uhcvuk7jhxd5cxrvh23xulkcewd8ssyf38ec"
+        assert parsed.contracts[0].owner_address == "erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th"
+
+    @pytest.mark.networkInteraction
```
Comment (andreibancioiu) [inline multiversx_sdk/core/transactions_outcome_parsers/smart_contract_transactions_outcome_parser_test.py:137]: :+1:

**Review by andreibancioiu — APPROVED** [review]

File: requirements.txt [hunk 6]
```diff
@@ -1,6 +1,6 @@
-pycryptodomex==3.16.0
-protobuf==3.20.1
-cryptography==36.0.2
+pycryptodomex==3.19.1
+protobuf==3.20.2
+cryptography==42.0.2
```
[the remark this answers is not in the document]
Comment (popenta) [reply requirements.txt:3]: Will take a look, and will also adjust the workflow in the next PR.

File: multiversx_sdk/converters/transactions_converter_test.py (hunk 2 above)
Comment (popenta) [reply multiversx_sdk/converters/transactions_converter_test.py:148]: No "native" way as far as I know.  Will think about doing a custom `deep_equal()` method to use in some tests.

File: multiversx_sdk/core/transactions_outcome_parsers/smart_contract_transactions_outcome_parser.py (hunk 3 above)
Comment (popenta) [reply multiversx_sdk/core/transactions_outcome_parsers/smart_contract_transactions_outcome_parser.py:25]: Indeed, I was thinking the same.

**Review by ssd04 — APPROVED** [review]

**Pull request merged**